packages/test/test-end-to-end-tests/src/test/dataMigration.spec.ts

```diff
@@ -0,0 +1,716 @@
+	summarizeNow,
+	waitForContainerConnection,
+} from "@fluidframework/test-utils";
+import {
+	BaseContainerRuntimeFactory,
+	ContainerRuntimeFactoryWithDefaultDataStore,
+	DataObject,
+	DataObjectFactory,
+	defaultRouteRequestHandler,
+} from "@fluidframework/aqueduct";
+import { SharedDirectory } from "@fluidframework/map";
+import {
+	ContainerRuntime,
+	DefaultSummaryConfiguration,
+	IContainerRuntimeOptions,
+	ISummarizer,
+	summarizerClientType,
+} from "@fluidframework/container-runtime";
+import { RuntimeFactoryHelper, requestFluidObject } from "@fluidframework/runtime-utils";
+import { IDocumentStorageService } from "@fluidframework/driver-definitions";
+import { FluidObject, IFluidHandle, ITelemetryBaseLogger } from "@fluidframework/core-interfaces";
+import { Deferred } from "@fluidframework/common-utils";
+import { IChannelFactory, IFluidDataStoreRuntime } from "@fluidframework/datastore-definitions";
+import {
+	IClientDetails,
+	IDocumentMessage,
+	IQuorumClients,
+	ISequencedDocumentMessage,
+	ISnapshotTree,
+	ISummaryAck,
+	ISummaryContent,
+	IVersion,
+	MessageType,
+	SummaryType,
+} from "@fluidframework/protocol-definitions";
+import { SharedString } from "@fluidframework/sequence";
+import {
+	FluidDataStoreRegistryEntry,
+	IFluidDataStoreRegistry,
+	NamedFluidDataStoreRegistryEntries,
+	NamedFluidDataStoreRegistryEntry,
+} from "@fluidframework/runtime-definitions";
+import { IContainerRuntime } from "@fluidframework/container-runtime-definitions";
+import { RuntimeRequestHandler } from "@fluidframework/request-handler";
+
+/**
+ * At the top level we are using the ContainerRuntimeFactoryManager which utilizes both the regular containerRuntimeFactory
+ * as well as a special containerRuntimeFactory (MigrationContainerRuntimeFactory) takes the regular registry and replaces all
+ * the factories with the MigrationDataObjectFactory. It also creates a MigrationContainerContext adapter and passes that
+ * to the new ContainerRuntime.
+ *
+ * The ContainerRuntimeFactoryManager detects when it's a summarizer client - we need to add logic to detect we are opening a v1
+ * form of the document, and loads the summarizer with the MigrationContainerRuntimeFactory.
+ *
+ * This makes sure all the dataObjects we load are MigrationDataObjects, which we control. Our customers can take this and extend
+ * or use it generically. I think we should leave it up to them.
+ *
+ * They then would make the changes on the summarizer client, maybe via op or something, not exactly sure. I think the main client
+ * should be the one to send the op, but that's a small detail. In my opinion I think the main client can do detection as that
+ * runtime already is established and it can communicate back to the runtime factory to tell it to create a migration summarizer.
+ *
+ * A thought I'm concerned about is that the search blob can disappear for all dataObjects that are modified, though search will
+ * resume on the next summarization (probably).
```

Comment:
Since the app implements the DataMigrationObject, perhaps the app can regenerate and include the search blob as part of the migration step?

Reply:
Yeah, I think they can definitely do that.

Reply:
I think the thing I worry about is that they have their own implementation of the fluid runtime, it would be nice if they didn't need to do that.

I think it would be better if we provided a better API where they just call into the datastore runtime to pregenerate the search blob